zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java

```diff
@@ -1927,4 +1927,46 @@ private void broadcastToWatchers(String noteId, String subject, Message message)
       }
     }
   }
+
+  @Override
+  public void onParaInfosReceived(String noteId, String paragraphId,
+      String interpreterSettingId, Map<String, String> metaInfos) {
+    Note note = notebook().getNote(noteId);
+    if (note != null) {
+      Paragraph paragraph = note.getParagraph(paragraphId);
+      if (paragraph != null) {
+        InterpreterSetting setting = notebook().getInterpreterFactory()
+                                               .get(interpreterSettingId);
+        setting.addNoteToPara(noteId, paragraphId);
+        metaInfos.remove("noteId");
+        metaInfos.remove("paraId");
+        String label = metaInfos.get("label");
+        metaInfos.remove("label");
+        paragraph.updateRuntimeInfos(label, metaInfos, setting.getGroup());
+        broadcast(note.getId(), new Message(OP.PARAGRAPH).put("paragraph", paragraph));
+      }
+    }
+  }
+
+  public void clearParagraphRuntimeInfo(InterpreterSetting setting) {
```

Comment:
I am afraid this method may some issues because the binding between `InterpreterSetting` and `Note/Paragraph` may change. e.g. At the beginning I use spark in p1, but later I use flink in p1, but the method will clear all the runtime info of p1 when I restart spark interpreter.